core/dbt/context/providers.py

```diff
@@ -212,16 +212,19 @@ def __call__(self, *args: str) -> Union[str, RelationProxy, MetricReference]:
 
 class BaseRefResolver(BaseResolver):
     @abc.abstractmethod
-    def resolve(self, name: str, package: Optional[str] = None) -> RelationProxy:
+    def resolve(
+        self, name: str, package: Optional[str] = None, version: Optional[str] = None
+    ) -> RelationProxy:
         ...
 
-    def _repack_args(self, name: str, package: Optional[str]) -> List[str]:
-        if package is None:
-            return [name]
-        else:
-            return [package, name]
+    def _repack_args(self, name: str, package: Optional[str], version: Optional[str]) -> List[str]:
+        repacked_args = [name] if package is None else [package, name]
+        if version:
+            repacked_args.append(f"version:{version}")
```

Comment:
Opted to repack the `version` arg as a formatted string to keep the structure of `model.refs` as `List[List[str]]` (result of this method is [appended to model.refs here](https://github.com/dbt-labs/dbt-core/pull/7204/files#diff-d14b7d00e52da5da96bfb3bdef414e9289d923efe0a9cdadd716d792c7199535R465). 

Leads to some [not-so-fun string parsing](https://github.com/dbt-labs/dbt-core/pull/7204/files#diff-a25a77362db78f4ddc0675bcbd4775e158631076c9918ce5b64cfcd9f6db6bceR1190) when refs are processed. 

Could also add it as a dictionary, which would hold all kwargs (if we ever add more) going forward. This would be a breaking change in the manifest (model.refs becomes `List[List[Union[str, Dict[str,str]]]]`, but would be more straightforward to parse when processing refs.

Reply:
Let's go with the dictionary. IMO:
- consumers of the manifest should use `depends_on.nodes`, rather than `refs`, in almost all cases
- the list-of-lists always felt a bit janky

This may require some changes in some of the weirder packages out there, but in a way that should make their lives easier, ultimately

Reply:
I'm good to go with a dictionary as well!

What I'd originally thought up in terms of representation in the manifest.json for `model.refs` was a `List[List[Union[str, Dict[str,str]]]]`. For example, 
```json
"refs": ["package_name", "model_name", {"version":"3", ...}]
```

But what might be best is actually just a `List[Dict[str, str]]`. For example, 
```json
"refs": {
"package": "package_name",  # what should we actually call this key..?
"model": "model_name",
"version": "version_name"
}
```

That would enable us to simplify how we parse `ref` [here](https://github.com/dbt-labs/dbt-core/pull/7204/files#diff-a25a77362db78f4ddc0675bcbd4775e158631076c9918ce5b64cfcd9f6db6bceR1190) as well.

Reply:
I vote for List[Dict[str, str]] too. It's always a bit messy to have two different types.  We're not allowing versioned sources, right?  Sources currently List[List[str]] but always have two strings in the list, whereas refs only have one. Eventually it might be nice to switch all of those things to dictionaries.

Reply:
> We're not allowing versioned sources, right?
right - agree it'd be nice to switch them all to dictionaries. I can make an issue for sources, perhaps we can tackle it sa follow-on tech debt work.